src/activity_handlers.cpp

```diff
@@ -293,10 +293,20 @@ void set_up_butchery( player_activity &act, player &u, butcher_type action )
             has_table_nearby = true;
         }
     }
+    bool has_tree_nearby = false;
+    for( const tripoint &pt : g->m.points_in_radius( u.pos(), 2 ) ) {
+        if( g->m.has_flag( "TREE", pt ) ) {
+            has_tree_nearby = true;
+        }
+    }
     // workshop butchery (full) prequisites
     if( action == BUTCHER_FULL ) {
-            if( !g->m.has_flag_furn( "BUTCHER_EQ", u.pos() ) ) {
-                u.add_msg_if_player( m_info, _( "You need a butchering rack to perform a full butchery." ) );
+            if( ( u.has_amount( "rope_30", 1 ) || u.has_amount( "rope_makeshift_30", 1 ) ) && !has_tree_nearby && !g->m.has_flag_furn( "BUTCHER_EQ", u.pos() ) ) {
+                u.add_msg_if_player( m_info, _( "You have a rope to lift the corpse but there is no tree nearby." ) );
+                act.set_to_null();
+                return;
+            } else if( !u.has_amount( "rope_30", 1 ) && !u.has_amount( "rope_makeshift_30", 1 ) && !g->m.has_flag_furn( "BUTCHER_EQ", u.pos() ) ) {
```

Comment:
Can hoist the `u.has_amount( string, count )` calls out of the conditions into the scope of `if( action == BUTCHER_FULL )`. Currently if the first condition fails it will have calculated if you have enough of `rope_30` or `rope_makeshift_30`, and then recalculates at least `rope_30`. Moving calculation out of the condition testing will always only calculate each once.
```
if ( action == BUTCHER_FULL ){
    bool has_rope_30 = u.has_amount( "rope_30", 1);
    bool has_makeshift_rope_30 = u.has_amount( "rope_makeshift_30", 1);

    if ( ( has_rope_30 || has_makeshift_rope_30 ) && ...){ ... }
    else if ( !has_rope_30 && !has_makeshift_rope_30 && ...){ ... }
[...]
```

Reply:
And also I can compile both bools into one and do the same with BUTCHER_EQ check. Thanks, good advice.